### DISCUSSION TITLE
Detach an OT device using Dataset API

### DISCUSSION THREAD

**Discussion by farhangj** [openthread/openthread#8107] [Q&A] [answered]
Hi,

I'm using otDatasetSetActive() to set the network params (NetKey, PAN ID, etc).

I have noticed that if devices are already on the same network, (i.e. given the same parameters), then if I want to isolate/detach one device from the network, by setting new params and creating a new network, using the same API, I cannot detach the device.
It seems like the other routers in the network end up receiving the same parameters and "move" to the new network.

I confirmed via CLI that "dataset pending" gets set as well to what I was hoping only be active dataset, which means that the router I am trying to detach is deliberately distributing the new key to other nodes. '
Nothing is mentioned in the [API of otDatasetSetActive()](https://openthread.io/reference/group/api-operational-dataset#otdatasetsetactive) about this.

My understanding was as long as OT Network Key is changed, it should be detached. but no!

How can I disable/enable this feature?

Thanks!

**Comment by jwhui** [comment]
If the device is attached to an existing Thread network, [`otDatasetSetActive()`](https://openthread.io/reference/group/api-operational-dataset#otdatasetsetactive) will attempt to set the Active Dataset on the Thread network, not just itself.

If you want to isolate the device itself, you can detach the device from the Thread network first, by calling [`otThreadSetEnabled()`](https://openthread.io/reference/group/api-thread-general#otthreadsetenabled).

See relevant code: https://github.com/openthread/openthread/blob/3d60038bcb65d02e93f747f81ff3994352492cb0/src/core/meshcop/dataset_manager.cpp#L193-L220

**Comment by abtink** [comment, marked as answer]
To detach specific device, we can call APIs on it to stop/detach directly:
- `otThreadSetEnabled()` to stop Thread MLE operation (device will detach immediately)
- or `otThreadDetachGracefully()` to do a graceful detach (inform other devices before going away).

Then to move the device to another network, first need to factory reset it (clear non-volatile settings):
- `otInstanceFactoryReset()` and/or `otInstanceErasePersistentInfo()` to erase old persisted dataset on device 

Then can add new dataset on it to start/join a new network.

----

The behavior of "active/pending dataset" is correct/intended. The way to think about this is that Operational Dataset belongs to a Thread mesh and the API is changing it on the Thread mesh.

**Reply by farhangj** [reply]
Thanks @abtink  for clarifying these point. This resolved the issue I was having.

**Reply by farhangj** [reply]
@abtink  in my version of OT source, I don't have access to otThreadDetachGracefully(), does otThreadBecomeDetached() do the same thing? except it won't notify other routers that it's leaving? What could be the downside of not leaving gracefully?

**Reply by jwhui** [reply]
With `otThreadBecomeDetached()`, the Thread device will attempt to reattach immediately after detaching.

With `otThreadSetEnabled()`, the Thread device will detach without notifying other devices and remain detached. Not notifying other devices will mean that any state for the device will remain in the Thread network until it times out. For example, a parent maintains state for its children and will buffer incoming messages for its children. If the child is a sleepy device, the parent will also respond to Address Query messages on behalf of the child.
